### ISSUE TITLE
Error when parsing complex f-string

### ISSUE THREAD

**Issue by twoertwein** [microsoft/pyright#2060]
```py
x = 0
precision = 3
print(f"{x: .{precision:d}f}")
```
error: Unexpected token at end of expression

This expression is used in pandas:
https://github.com/pandas-dev/pandas/blob/dad3e7fc3a2a75ba5f330899be0639cff0f73f6c/pandas/io/formats/format.py#L1280

**Comment by erictraut** [comment]
PEP 498 doesn't mention that it's legal to use a colon token within an expression statement within a format specifier. I'm not quite sure what this even means, but it appears that it's accepted by the Python interpreter. I've updated the pyright code to deal with this condition.

Thanks for reporting the issue.

**Comment by jakebailey** [comment]
It's here: https://www.python.org/dev/peps/pep-0498/#format-specifiers

After the `:`, effectively anything is legal, as it's just a string passed to a `__format__` method. (This did already work in my f-string implementation in MPLS, it seems, though now that I am checking our test files, Pylance/pyright don't correctly handle `yield` in fstrings, another unrelated problem.)

**Comment by erictraut** [comment]
I appreciate that anything after the `:` represents the format specification, and any string is permitted. However, in this case, the expression in question is the format specification. I'm not sure what it means to include a format specification (`:d`) for an expression used within a format specification (`: .{precision:d}f`).

**Comment by jakebailey** [comment]
I guess I'm not sure what you mean; this:

```python
f"{x: .{precision:d}f}"
```

May as well be:

```python
f"{x:any random string $#@!%% hello}"
```

Since it's the first colon that matters. It's entirely up to `x.__format__` to then go do something with that string and we don't need to check validity.

**Comment by jakebailey** [comment]
Oh, my mistake, that gets stuck in there early. I'll have to reread.


```python
class Foo:
    def __format__(self, x):
        print(x)
        return "Foo"
        
x = Foo()
precision = 3
print(f"{x: .{precision:d}f}")
```

prints

```
 .3f
Foo
```

**Comment by erictraut** [comment]
The format string can contain dynamic expressions that are evaluated by the Python interpreter. These expressions must be enclosed in curly braces. Pyright needs to know about these because it needs to evaluate these expressions (to validate their validity and mark the any symbols as referenced).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.156, which I just published. It was already included in the last release of pylance.
